src/cdadrv.c

```diff
@@ -1,405 +1,405 @@
-		goto err_cls_reg;
-
-	if( (req_pci_did || req_pci_vid) && pci_id_table_size >= 2 ) {
-		// Last table element is 0,0
-		// Update pre-last item
-		cda_pci_ids[pci_id_table_size-2].vendor = req_pci_vid;
-		cda_pci_ids[pci_id_table_size-2].device = req_pci_did;
-	}
-	ret = pci_register_driver(&cda_pci);
-	if( ret )
-        goto err_pci_reg_drv;
-
-	return 0;
-
-err_pci_reg_drv:
-	class_unregister(&cda_class);
-err_cls_reg:
-	unregister_chrdev_region(cdadev_first, CDA_DEV_MINOR_MAX);
-err_alloc_cdev_reg:
-	return ret;
-}
-
-static void __exit dcadrv_exit(void)
-{	
-	if( test_probe ) { 
-		printk("Stop test run. Nothing initialized\n");
-		return;
-	}
-    pci_unregister_driver(&cda_pci);
-	class_unregister(&cda_class);
-	unregister_chrdev_region(cdadev_first, CDA_DEV_MINOR_MAX);
-}
-
-module_init(cdadrv_init);
-module_exit(dcadrv_exit);
+// SPDX-License-Identifier: GPL-2.0
+// Copyright(c) 2020 DeGirum Corp., Egor Pomozov.
+//
+// CDA linux driver mem blocks/mem maps and interrupt request handler
+//
+// This program is free software; you can redistribute it and/or modify it
+// under the terms and conditions of the GNU General Public License,
+// version 2, as published by the Free Software Foundation.
+//
+#include <linux/module.h>
+#include <linux/errno.h>
+#include <linux/init.h>
+#include <linux/fs.h>
+
+#include "cdadrv.h"
+#include "cdaioctl.h"
+
+MODULE_AUTHOR("DeGirum Corp., Egor Pomozov");
+MODULE_DESCRIPTION("CDA linux driver to access pci devices");
+MODULE_LICENSE("GPL");
+MODULE_VERSION("0.5.0.3");
+// The version has to be in the format n.n.n.n, where each n is a single digit
+
+#if LINUX_VERSION_CODE < KERNEL_VERSION(4,9,0)
+#error Too old kernel
+#endif
+
+static dev_t cdadev_first;
+static const char cda_name[] = "cda";
+static int req_pci_did = 0;
+static int req_pci_vid = 0;
+static int test_probe = 0;
+
+#define CDA_DEV_MINOR_MAX 32
+static DEFINE_SPINLOCK(cdadevlist_sl);
+static DEFINE_IDA(cdaminor_ida);
+static LIST_HEAD(cdadevs);
+
+// Module parameters
+module_param_named(did, req_pci_did, int, 0644);
+MODULE_PARM_DESC(did, "Set required PCI device ID");
+module_param_named(vid, req_pci_vid, int, 0644);
+MODULE_PARM_DESC(vid, "Set required PCI vendor ID");
+module_param_named(test_probe, test_probe, int, 0644);
+MODULE_PARM_DESC(test_probe, "Check permissions to load driver");
+
+static void cdadev_release(struct device *kdev);
+static void cda_pci_remove(struct pci_dev *pcidev);
+static int cda_pci_probe(struct pci_dev *pcidev,
+			       const struct pci_device_id *id);
+
+static int cda_cdev_open(struct inode *ino, struct file *file);
+static int cda_cdev_release(struct inode *ino, struct file *file);
+static long cda_cdev_ioctl(struct file *file, unsigned int cmd, unsigned long arg);
+
+static struct pci_device_id cda_pci_ids[] = {
+	{ PCI_DEVICE(0x1f0d, 0x0100) },
+	{ PCI_DEVICE(0x1f0d, 0x8101) },
+	{ PCI_DEVICE(0x1f0d, 0x0101) },
+	{ PCI_DEVICE(0x10ee, 0x8011) },
+	{ }
+};
+MODULE_DEVICE_TABLE(pci, cda_pci_ids);
```

Comment:
Empty lines are required to support the hardware which is not released yet.

Reply:
Ah, I see, one is used that way. That's still wrong, but I'll split it into two commits then.